tests/main/views/test_buyers.py

```diff
@@ -3676,3 +3642,122 @@ def test_award_brief_details_post_raises_400_and_displays_error_messages_and_pre
+
+        resp = self.client.get(self.url.format(brief_id=self.brief["id"]))
+
+        assert resp.status_code == 404
+
+    @pytest.mark.parametrize('status', ['withdrawn', 'draft', 'live', 'cancelled', 'unsuccessful', 'awarded'])
+    def test_404_if_brief_not_closed(self, status):
+        self.data_api_client.get_brief.return_value['briefs']['status'] = status
+
+        resp = self.client.get(self.url.format(brief_id=self.brief["id"]))
+
+        assert resp.status_code == 404
+
+    @pytest.mark.parametrize('framework_status', ['coming', 'open', 'pending', 'standstill'])
+    def test_404_if_incorrect_framework_status(self, framework_status):
+        self.data_api_client.get_framework.return_value['frameworks']['status'] = framework_status
+
+        resp = self.client.get(self.url.format(brief_id=123))
+
+        assert resp.status_code == 404
+
+    def test_that_no_option_chosen_triggers_error(self):
+        resp = self.client.post(self.url.format(brief_id=123))
+        document = html.fromstring(resp.get_data(as_text=True))
+        validation_message = document.xpath('//span[@class="validation-message"]')[0].text_content()
+        assert resp.status_code == 400
+        assert "Answer required" in validation_message
+
+    def test_cancel_triggers_cancel_brief(self):
+        resp = self.client.post(
+            self.url.format(brief_id=123), data={'cancel_reason': 'cancel'}
+        )
+
+        assert resp.status_code == 302
+        self.data_api_client.cancel_brief.assert_called_once_with('123', user='buyer@email.com')
+
+    def test_unsuccessful_triggers_cancel_brief(self):
+        resp = self.client.post(
+            self.url.format(brief_id=123), data={'cancel_reason': 'unsuccessful'}
+        )
+
+        assert resp.status_code == 302
+        self.data_api_client.update_brief_as_unsuccessful.assert_called_once_with('123', user='buyer@email.com')
+
+    @pytest.mark.parametrize('status', ['withdrawn', 'draft', 'live', 'closed', 'awarded'])
+    def test_400_if_incorrect_status_supplied(self, status):
+        resp = self.client.post(
+            self.url.format(brief_id=123), data={'cancel_reason': status}
+        )
+
+        assert resp.status_code == 400
+        assert "Not a valid choice" in resp.get_data(as_text=True)
+        assert self.data_api_client.update_brief_as_unsuccessful.called is False
+        assert self.data_api_client.cancel_brief.called is False
+
+    @pytest.mark.parametrize('status', ['cancel', 'unsuccessful'])
+    def test_redirect_on_successful_status_change(self, status):
+        resp = self.client.post(
+            self.url.format(brief_id=123), data={'cancel_reason': status}
+        )
+
+        redirect_text = html.fromstring(resp.get_data(as_text=True)).text_content().strip()
+        expected_url = '/buyers/frameworks/digital-outcomes-and-specialists-2/requirements/digital-outcomes/123'
+
+        assert resp.status_code == 302
+        assert expected_url in redirect_text
```

Comment:
Can we assert here whether the flash message is displayed (as we do in the awards test class - `self.assert_flashes("updated-brief")`)?

Reply:
Yes please

Reply:
👍